[Wong Rende] iP

WR3nd3 opened the pull request:
## Duke
> "Thus, a 4-MC module would require 10 hours of work a week, including lectures, tutorials, laboratory sessions, assignments, and independent or group study." – [NUS Registrar's Office](https://www.nus.edu.sg/registrar/academic-information-policies/graduate/modular-system)

Duke *frees* your mind of having to choose what you need to do. It
1.  Solves your dilemma of whether to work or sleep.
     - Choice is an illusion
2. Demands more OOP
     - OOP is love, OOP is life
3.  Everything will turn out fine.
     - You cannot stop it.

And it is **FREE!**

Features:
- [x] Blended Learning
- [x] Sleep Deprivation
- [x] Cats 🐱 

If you submit to our feline overlords, you can print them too. Here's the 
```
private static final String CAT_SYMBOL = "     /\\_____/\\\n"
          + "    /  o   o  \\\n"
          + "   ( ==  ^  == )\n"
          + "    )         (\n"
          + "   (           )\n"
          + "  ( (  )   (  ) )\n"
          + " (__(__)___(__)__)\n"
          + "           _\n"
          + "  ___ __ _| |_ ___\n"
          + " / __/ _` | __/ __|\n"
          + "| (_| (_| | |_\\__ \\\n"
          + " \\___\\__,_|\\__|___/\n";
```

Commit:
Add Gradle support

WR3nd3 committed:
Replaced Duke with Cat logo.

WR3nd3 committed:
Added Recorder class to implement echo and exit function.

WR3nd3 committed:
Completed Level-2 by adding a ToDoList and Task class. Commented out code for Echo functionality.

WR3nd3 committed:
Completed Level 3 and changed  control flow to be controlled by enum and switch with a try-catch block.

WR3nd3 committed:
Changed Task to abstract class that is a superclass to the new Todo, Event, and Deadline classes. Added new enum keywords.

WR3nd3 committed:
Completed Level-4 by editing the control flow and enum keywords alongside minor format issues.

WR3nd3 committed:
Finished Levvel-6. Remaining tasks include clean up, automated testing, and documentation. Error handling for level-5 is covered to an extent in a try-catch block for all invlalid entries for all commands

WR3nd3 committed:
Completed A-TextUiTesting and added documentation.

WR3nd3 committed:
Minor style fixes

WR3nd3 committed:
Add ListLoader class and modified other classes to accomodate its behaviour.

WR3nd3 committed:
Add ListLoader class methods to modify state of duke.txt. Bug with MarkTask.

WR3nd3 committed:
Add appendToList, markTask, unmarkTask and delete methods to ListLoader class. Save functionality is complete.

WR3nd3 committed:
Commit duke.txt.

WR3nd3 committed:
Add date and time recognition features to Event and Deadline classes.

WR3nd3 committed:
Merge branch 'branch-Level-7'

WR3nd3 committed:
Merge branch 'branch-Level-8'

WR3nd3 committed:
Fix ListLoader imports.

WR3nd3 committed:
Add more OOP through command subclasses, and separate classes for ui and parsing. Neatening and java docs required.

WR3nd3 committed:
Add additional abstraction to the classes alongside javadoc comments.

WR3nd3 committed:
Packaged all classes.

WR3nd3 committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip

WR3nd3 committed:
Add Gradle.

WR3nd3 committed:
Add JUnit tests for Deadline and Parser classes.

WR3nd3 committed:
Add JAR file.

WR3nd3 committed:
Add edit to TaskId javadoc.

WR3nd3 committed:
Tweak to coding standard

Static testing through checkstyle revealed multiple coding style
violations.

Let's fix these violations.

WR3nd3 committed:
Add Find functionality

Requested to add find functionality as a command to print list of
tasks with descriptions matching an entered string.

FindCommand class created and supporting methods added to Task,
TaskList and Ui classes.

Let's improve the javadoc as well.

WR3nd3 committed:
Merge branch 'branch-Level-9'

WR3nd3 committed:
Merge branch 'branch-A-CodingStandard'

WR3nd3 committed:
Merge branch 'branch-A-JavaDoc'

JasonYapzx commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,217 @@
+            + "  ___ __ _| |_ ___\n"
+            + " / __/ _` | __/ __|\n"
+            + "| (_| (_| | |_\\__ \\\n"
+            + " \\___\\__,_|\\__|___/\n";
```
I like Cat art and the personalised style of the Duke helper :)

JasonYapzx commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,85 @@
+    protected boolean isCompleted;
+    private String addOn = "nya!";
+    private String notDoneSymbol = "[Zzzzzzz]";
+    private String doneSymbol = "[=^._.^=]";
```
Great to see that all the class variables are encapsulated and never declared public.

JasonYapzx commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,85 @@
+     */
+    public boolean contains(String content) {
+        return description.contains(content);
+    }
```
Would using `hasContent` instead of `contains` as a boolean name be more in line with the coding standard?

WR3nd3 replied:
Fixed, thanks!

JasonYapzx commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,217 @@
+    private static final String SERVICE = "What can I do for mew?\n";
+
+    /** Goodbye message printed when exiting duke.Duke */
+    private static final String GOODBYE = "Bye! See nya later!\n";
```
Should these have a common prefix? For example:
`MESSAGE_BORDER`, `MESSAGE_SERVICE`, `MESSAGE_GOODBYE`, since these constant messages can be grouped together?

WR3nd3 replied:
Fixed, thanks!

JasonYapzx reviewed (commented):
LGTM to me, just some small coding standard nits to pick. Well done 💯 👍🏻

WR3nd3 committed:
Fix coding quality

Based on PR review, naming of variables should change in Ui class and
naming of methods should change in Task class.

Scope of variables in Task class is also narrowed by changing global
variables to local variables.

WR3nd3 committed:
Add GUI support to Duke

According to project specifications, GUI support was enabled via
JavaFx.

To accommodate the change, methods in Ui and Command and Duke classes
are modified.

Javadoc is completed. The Duke program works in both CLI and GUI form
now.

WR3nd3 committed:
Add GUI support to Duke

According to project specifications, GUI support was enabled via
JavaFx.

To accommodate the change, methods in Ui and Command and Duke classes
are modified.

Javadoc is completed. The Duke program works in both CLI and GUI form
now.

WR3nd3 committed:
Merge branch 'branch-Level-10'

WR3nd3 committed:
Add welcome message for GUI version of Duke

Duke could not show the welcome message and symbol on GUI before.
Additional methods were added to duke and MainWindow classes to
facilitate this.

WR3nd3 committed:
Adhere to checkstyle

nitant-p commented on src/main/java/duke/storage/ListLoader.java:
```diff
@@ -0,0 +1,281 @@
+        BufferedReader reader = null;
+        try {
+            // Create necessary save file if it does not exist
+            if (!new File("data").exists()) {
```
Maybe just create the File object once (before the if statement) and then do the checks instead of two separate objects?

nitant-p commented on src/main/java/duke/storage/ListLoader.java:
```diff
@@ -0,0 +1,281 @@
+                Boolean isCompleted = inputArray[1].equals("1") ? true : false;
+                String description = inputArray[2];
+
+                switch (TaskId.valueOf(command)) {
```
This Switch-Case block is super simple and easy to follow! 😄

WR3nd3 replied:
Thank you

nitant-p commented on src/main/java/duke/exception/DukeException.java:
```diff
@@ -0,0 +1,54 @@
+    private static final String TAB = "    ";
+
+    /** Default error message for invalid commands */
+    private static final String defaultMsg = INTRO
```
I think this constant should be called DEFAULT_MSG instead 🤔

nitant-p commented on src/main/java/duke/exception/DukeException.java:
```diff
@@ -0,0 +1,54 @@
+    private static final String LIST = "Input 'list' for overview\n";
+    private static final String BYE = "Input 'bye' to exit.\n";
+    private static final String OUTRO = "NYAAAAAA!\n";
+    private static final String TAB = "    ";
```
What about using \t instead of the empty spaces for formatting? 🤔

WR3nd3 replied:
Oo I didn't know about that, thanks

nitant-p commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,80 @@
+     * @return a string consisting of the deadline completion status and description.
+     */
+    @Override
+    public String toString() {
```
Maybe put the entire if-else block into a separate method so that all these checks only have to executed once and not every time toString() is called 🤔

nitant-p reviewed (commented):
I think your program is super detailed yet the readability and code quality is quite good so it wasn't difficult to follow! Actually learned quite a few things after reviewing your code 😄

WR3nd3 committed:
Add assertions

Parser :: parse takes in a String as input however it does not
account for the scenario of being passed a null object.

Although this should not occur, using assertions to verify that the
input is not null would grant assurance and safety.

Let's add an assertion to Parser::parse to prevent runtime exceptions.

WR3nd3 committed:
Merge pull request #1 from WR3nd3/branch-assertion

Add assertions

WR3nd3 committed:
Fix code quality

The recent PR review revealed areas of code where code quality can be
improved.

Following the advice given, the following changes were made to improve
code quality:
* Renaming of default_msg constant in DukeException via refactoring.
* Abstraction of string formatter for dates and time within the
  toString() method of Event and Deadline classes.
* Removing redundant constant TAB in Ui class in favour of using "/t"
  where necessary.

WR3nd3 committed:
Merge pull request #2 from WR3nd3/branch-A-CodeQuality

Fix code quality

WR3nd3 committed:
Add streams

Stream methods are useful to increase clarity in code.

Let's modify giveFindList and giveList in TaskList and their
corresponding Ui methods to use streams.

WR3nd3 committed:
Merge branch 'master' into branch-A-Streams

WR3nd3 committed:
Merge pull request #3 from WR3nd3/branch-A-Streams

Add streams

WR3nd3 committed:
Merge branch 'branch-A-Assertions'

WR3nd3 committed:
Merge branch 'branch-A-CodeQuality'

WR3nd3 committed:
Merge branch 'master' of https://github.com/WR3nd3/ip

* 'master' of https://github.com/WR3nd3/ip:
  Add streams

WR3nd3 committed:
Add C-Help

Users may not know what commands are available.

The HelpCommand is added and invoked to show a list of valid commands
whenever help is inputted or if an invalid command is used.

Additionally, the GUI app now exits on Bye.

WR3nd3 committed:
Fix GUI box sizing

WR3nd3 committed:
Improve GUI and code quality

The GUI interface cannot resize and the text does not wrap around
properly.

The parse method in Parser is long and could become unwieldy for future
iterations and updates.

Let's,
* Update the properties of dialogbox.fxml in scenebuilder
* Rename the application to ManMeowth for personality
* Extract methods from within the parse method

WR3nd3 committed:
Add more abstraction

ListLoader class has long methods due to repeated boiler plate code
on reading and writing to the file.

This reduces readability.

Let's abstract out the segments of code for reading and writing into
their own separate methods.

WR3nd3 committed:
Merge branch 'listloader'

* listloader:
  Add more abstraction

WR3nd3 committed:
no message

WR3nd3 committed:
Add user guide

A brief user guide is not included.

Let's add a user guide in the README to guide users.

WR3nd3 committed:
Edit user guide

WR3nd3 committed:
Fix javadocs
